test/gtest/type-builder.cpp

```diff
@@ -661,3 +661,63 @@ TEST_F(NominalTest, TestDepth) {
   EXPECT_EQ(HeapType(HeapType::nofunc).getDepth(), size_t(-1));
   EXPECT_EQ(HeapType(HeapType::noext).getDepth(), size_t(-1));
 }
+
+// Test .traverseSubTypes() helper.
+TEST_F(NominalTest, TestTraverse) {
+  /*
+        A
+       / \
+      B   C
+           \
+            D
+  */
+  HeapType A, B, C, D;
+  {
+    TypeBuilder builder(4);
+    builder[0] = Struct();
+    builder[1] = Struct();
+    builder[2] = Struct();
+    builder[3] = Struct();
+    builder.setSubType(1, builder.getTempHeapType(0));
+    builder.setSubType(2, builder.getTempHeapType(0));
+    builder.setSubType(3, builder.getTempHeapType(2));
+    auto result = builder.build();
+    ASSERT_TRUE(result);
+    auto built = *result;
+    A = built[0];
+    B = built[1];
+    C = built[2];
+    D = built[3];
+  }
+
+  SubTypes subTypes({A, B, C, D});
+
+  using TypeSet = std::unordered_set<HeapType>;
+
+  auto getSubTypes = [&](HeapType type, Index depth) {
+    TypeSet ret;
+    subTypes.traverseSubTypes(
+      type, depth, [&](HeapType subType, Index depth) { ret.insert(subType); });
+    return ret;
+  };
+
+  EXPECT_EQ(getSubTypes(A, 0), TypeSet({A}));
+  EXPECT_EQ(getSubTypes(A, 1), TypeSet({A, B, C}));
+  EXPECT_EQ(getSubTypes(A, 2), TypeSet({A, B, C, D}));
+  EXPECT_EQ(getSubTypes(A, 3), TypeSet({A, B, C, D}));
+
+  EXPECT_EQ(getSubTypes(C, 0), TypeSet({C}));
+  EXPECT_EQ(getSubTypes(C, 1), TypeSet({C, D}));
+  EXPECT_EQ(getSubTypes(C, 2), TypeSet({C, D}));
+
+  // Manually test the depth parameter
```

Comment:
This wouldn't be difficult to roll into the previous tests by using a small struct with an `operator==` rather than `HeapType` in the `TypeSet`.

Reply:
Sorry, I don't understand?

Reply:
I guess it would be simpler to use `std::pair`. Here's what I have in mind for checking the subtypes and their depths at the same time:

```c++
using TypeSet = std::unordered_map<std::pair<HeapType, Index>>;
...
EXPECT_EQ(getSubTypes(A, 0), TypeSet({{A, 0}}));
...
```

Reply:
Ah, now I see, thanks! Yes, that is better. Done.